client/network-gossip/src/bridge.rs

```diff
@@ -147,16 +168,40 @@ impl<B: BlockT> Future for GossipEngine<B> {
 						this.state_machine.peer_disconnected(&mut *this.network, remote);
 					},
 					Event::NotificationsReceived { remote, messages } => {
-						let engine_id = this.engine_id.clone();
-						this.state_machine.on_incoming(
+						let messages = messages.into_iter().filter_map(|(engine, data)| {
+							if engine == this.engine_id {
+								Some(data.to_vec())
+							} else {
+								None
+							}
+						}).collect();
+
+						let to_forward = this.state_machine.on_incoming(
 							&mut *this.network,
 							remote,
-							messages.into_iter()
-								.filter_map(|(engine, data)| if engine == engine_id {
-									Some(data.to_vec())
-								} else { None })
-								.collect()
+							messages,
 						);
+
+						for (topic, notification) in to_forward.into_iter() {
+							if let Entry::Occupied(mut entry) = this.message_sinks.entry(topic) {
+								trace!(
+									target: "gossip",
+									"Pushing consensus message to sinks for {}.", topic,
+								);
+								entry.get_mut().retain(move |sink| {
+									if let Err(e) = sink.unbounded_send(notification.clone()) {
```

Comment:
With this pull request sending on the event channel is happening in an asynchronous context. Thus we could replace the unbounded channel with a bounded channel in a follow up pull request.